pkg/generate/code/equal.go

```diff
@@ -0,0 +1,236 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	awssdkmodel "github.com/aws/aws-sdk-go/private/model/api"
+
+	ackmodel "github.com/aws-controllers-k8s/code-generator/pkg/model"
+	"github.com/aws-controllers-k8s/code-generator/pkg/names"
+)
+
+// CheckEqualType returns Go code that checks the equality of two struct types.
+// The generated code returns false at the first spotted difference.
+func CheckEqualType(
+	typeDef *ackmodel.TypeDef,
+	// String representing the name of the variable that represents the first
+	// object under comparison. This will typically be something like "a"
+	firstVarName string,
+	// String representing the name of the variable that represents the second
+	// object under comparison. This will typically be something like "b".
+	secondVarName string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	out := ""
+
+	// We need a deterministic order to loop over attributes
+	attrsNames := []string{}
+	for attrName := range typeDef.Attrs {
+		attrsNames = append(attrsNames, attrName)
+	}
+	sort.Strings(attrsNames)
+
+	// For each attribute generate the equality check logic
+	for _, attrName := range attrsNames {
+		attr := typeDef.Attrs[attrName]
+		out += checkEqualField(attrName, attr.Shape, firstVarName, secondVarName, indentLevel)
+	}
+	return out
+}
+
+// equalStruct outputs Go code that compares two similar fields from different objects.
```

Comment:
documentation should start with method name "checkEqualField", correct?

Reply:
Yep ++